src/parser/metrics/RealLinesOfCode.ts

```diff
@@ -16,75 +12,70 @@ let dlog: DebugLoggerFunction = debuglog("metric-gardener", (logger) => {
  * Counts the number of lines in a file, not counting for comments and empty lines.
  */
 export class RealLinesOfCode implements Metric {
-    private commentStatementsSuperSet: QueryStatementInterface[] = [];
+    private commentStatementsSet: Set<string>;
 
     /**
      * Constructs a new instance of {@link RealLinesOfCode}.
      * @param allNodeTypes List of all configured syntax node types.
      */
     constructor(allNodeTypes: ExpressionMetricMapping[]) {
-        allNodeTypes.forEach((expressionMapping) => {
-            if (
-                expressionMapping.metrics.includes(this.getName()) &&
-                expressionMapping.type === "statement"
-            ) {
-                const { expression, activated_for_languages } = expressionMapping;
-                const queryStatement = new ExpressionQueryStatement(
-                    expression,
-                    activated_for_languages
-                );
+        this.commentStatementsSet = new Set(
+            getExpressionsByCategory(allNodeTypes, this.getName(), "comment")
+        );
+    }
 
-                if (expressionMapping.category === "comment") {
-                    this.commentStatementsSuperSet.push(queryStatement);
-                }
-            }
-        });
+    /**
+     * Waling through the tree in order to find actual code lines.
+     *
+     * Uses a {@link TreeCursor} for this, as according to the
+     * {@link https://tree-sitter.github.io/tree-sitter/using-parsers#walking-trees-with-tree-cursors|Tree-sitter documentation},
+     * this is the most efficient way to traverse a syntax tree.
+     * @param cursor A {@link TreeCursor} for the syntax tree.
+     * @param sureCodeLines A set in which the line numbers of the found code lines are stored.
+     */
+    walkTree(cursor: TreeCursor, sureCodeLines = new Set<number>()) {
+        // This is not a comment syntax node, so assume it includes "real code".
+        if (!this.commentStatementsSet.has(cursor.currentNode.type)) {
+            // Assume that first and last line of whatever kind of node this is, is a real code line.
+            // This assumption should hold for all kinds of block/composed statements in (hopefully) all languages.
+            sureCodeLines.add(cursor.startPosition.row);
+            // sureCodeLines.add(cursor.endPosition.row);
```

Comment:
It does work as expected. I have tried quite hard to break it in the unit tests with the file resources/typescript/weird-lines.ts, but it does indeed also work with code like this:

````ts
for (i =
          0

     ;
      i < 100;
      i++) {
     // do something
     let a =
         5;
     a++ // a gets incremented
     ;
}

do {
     // do something
} while (x < /* x has to stay lower than 100 */ 100);
````

Adding all lines between start and end of each syntax nodes would
include thinks like empty lines inside code blocks. Or, in case of the
root node, even all lines of the whole file, so that does not work.

I first thought that I need to include at least the last row of a syntax node (end of a code block, etc.). However, as this approach traverses the concrete syntax tree, also including  any kind of "unnamed" syntax nodes like "{" or "(", any node that represents syntax that could mark the end of such a block is also visited, whether it being some kind of bracket or a last statement of a block (like in python).
This code would fail if there is some kind of syntax node that a) spans over multiple non-empty lines and b) does so without having child nodes in the syntax tree that start on these other, non-empty lines. I actually do not think that such a node exists (other than the "comment"-nodes, which should not be counted), as every non whitespace/non-control-character token of code that is relevant for the semantic of the program ("real code") seems to be included in the syntax tree, and I have no idea of a token of code that could span over multiple lines, as long as it is not allowed to have, for example, variable names like `"hello\nworld"` in a language.

TLDR: This surprisingly simple implementation "just works" :wink:

Reply:
Thinking of it, I could check for each node if it is a leaf node and if it does span over more than one line to handle such a case. Not sure if that would be worth the extra computing expense.

Reply:
It is a great implementation! I can't think of anything going wrong either. Let's work with this for now and if we find cases that need a different handling, we can still change it.

Reply:
Good. I will merge it once this PR is approved.